### ISSUE TITLE
Installed package won't import in notebook

### ISSUE THREAD

**Issue by fonnesbeck** [jupyter/notebook#2359]
I'm using the notebook from a conda env, and have the environment set up with all of my required packages installed. However, when I try to import one of these (seaborn) in the notebook, I get an import error suggesting it is not installed. This works fine from the command line and I can confirm using the `conda` command within the notebook that the package is installed:

![](http://d.pr/i/rkyJ+)

Running notebook 4.4.1 from conda.

Interestingly, when I try to import `notebook` from inside of a notebook, I get the same error:

```
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-2-09fdb2483669> in <module>()
----> 1 import notebook
      2 
      3 notebook.__version__

ImportError: No module named 'notebook'
```

**Comment by fonnesbeck** [comment]
To followup, I have even created a fresh conda env, but I get the exact same error.

**Comment by fonnesbeck** [comment]
I can further confirm that installing Seaborn via pip does not fix this issue.

**Comment by fonnesbeck** [comment]
Complete reinstallation of Anaconda fixed the issue.

**Comment by fonnesbeck** [comment]
This occurs with pip-installed packages as well as conda:

![](http://d.pr/i/6hhK+)

**Comment by ccordoba12** [comment]
I think you need to restart the kernel for this to work. Besides, you have to be sure to be using the right `pip` (e.g. `pip` that comes with Anaconda instead of system `pip`).

**Comment by fonnesbeck** [comment]
I've restarted the kernel multiple times, without success. I can also confirm that it is the pip installed into the env. I can import this library from the ipython console without issue.

**Comment by fonnesbeck** [comment]
Just to clarify, I am using a conda env based on this yml:

```yml
name: dev

channels:
 - conda-forge
 
dependencies:
- python=3.6
- arrow
- basemap
- blaze
- bokeh
- cython
- dask
- engarde
- ipython
- jupyter
- notebook
- ipyparallel
- ipywidgets
- joblib
- jupyter_client
- jupyter_core
- jupyter_contrib_nbextensions
- jupyter_nbextensions_configurator
- jupyterlab
- line_profiler
- mkl-service
- numexpr
- numba
- numpy
- openpyxl
- pandas
- patsy
- pillow
- pip
- pylint
- scikit-learn
- scipy
- seaborn
- snakeviz
- sympy
- tensorflow
- theano
- xlrd
- pip:
  - GPy
  - PyCap
  - pystan
  - rpy2
  - git+https://github.com/GPflow/GPflow.git#egg=GPflow
  - git+https://github.com/pymc-devs/pymc3.git#egg=pymc3
```

The env was created and all packages were installed without error, but some packages can be loaded into the notebook, while others were not.

**Comment by fonnesbeck** [comment]
Even installing the old fashioned way with `python setup.py install` does not result in the notebook seeing the package. I'm at a loss here.

**Comment by fonnesbeck** [comment]
Tried installing on an AWS instance. Same problem.

**Comment by ccordoba12** [comment]
What if you use a proper terminal instead of trying to run `!pip` in the notebook?

**Comment by fonnesbeck** [comment]
I was only using pip in the notebook to list the installed package. It was installed first with conda when I created the env and then using pip from a conventional terminal.

**Comment by fonnesbeck** [comment]
I can also confirm that the package is present in the `site-packages` directory for the active env:

```
drwxr-xr-x   24 fonnescj  staff      816 Apr  2 22:17 pymc3
drwxr-xr-x    7 fonnescj  staff      238 Apr  2 22:17 pymc3-3.0-py3.5.egg-info
-rw-r--r--    1 fonnescj  staff  1124023 Apr  2 14:30 pymc3-3.1rc3-py3.5.egg
```

And again, it imports from the IPython console; its only the notebook where it fails.

**Comment by fonnesbeck** [comment]
Steps to replicate:

1. Create conda env with:

    conda create --name tester notebook pymc3 numpy ipython

2. Activate environment:

    source activate tester

3. Open IPython, and import pymc3 from the command line. Assert that it imports as expected.

4. Open a Jupyter notebook session:

    jupyter notebook

5. Create a new Python3 notebook and import pymc3. Failure occurs:

```
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
<ipython-input-1-37bf2a3357ff> in <module>()
----> 1 import pymc3

ModuleNotFoundError: No module named 'pymc3'
```

**Comment by fonnesbeck** [comment]
I am able to get pymc3 to import if I install it in the root environment, but not in any environments that I create.

**Comment by takluyver** [comment]
Check `sys.executable` and `sys.path` inside the notebook - I bet your kernel is not running in the environment you intended it to. Run `jupyter kernelspec list` to see where it finds the `python3` kernel.

**Comment by fonnesbeck** [comment]
Wow, that's bizarre. Its an extremely odd list. I've deleted `~/Library/Jupyter/kernels` and that appears to have straightened things out. 

Thanks.

**Comment by mgooty** [comment]
@fonnesbeck Even I am facing the similar issue. I have installed keras but I am not able to import it in the notebook. I am using Mac. Where can I find `~/Library/Jupyter/kernels` so that I can follow your strategy?

**Comment by julienschroder** [comment]
Facing the same issue
@mgooty `!jupyter kernelspec list` will display the path but that didn't solve the issue for me.
Maybe more of a conda question

**Comment by lintusj1** [comment]
Also facing the same issue

**Comment by horsto** [comment]
Having the same issue, any ideas / solutions?

**Comment by takluyver** [comment]
It sounds like you're trying an import in two different installations of Python, or two different environments. Check `sys.executable` to see which Python and environment you're running in, and `sys.path` to see where it looks to import modules.

Python packages must be installed separately for each copy of Python you use, and if you are using virtualenvs or conda envs, packages must be installed into each environment where you need them. Either the package is not installed in one, or a different version of the package is installed.

To install packages to a particular Python installation or environment using pip, you can run it like this:

```shell
path/to/python -m pip install ...
# Replace path/to/python with the path of the Python executable (sys.executable)
```

If you are using conda environments, you can install packages from conda like this:

```shell
source activate myenv  # On Windows, just 'activate myenv' (no 'source')
conda install ...
```

I have written this up as a blog post as well: http://takluyver.github.io/posts/i-cant-import-it.html

[ This is a saved reply because I answer similar questions often - sorry if it doesn't exactly fit your case ]

**Comment by horsto** [comment]
Thanks for the tips! However: I installed jupyterlab in its own conda env and installed everything in there via conda install ... so it should be there, correct? Still - I cannot import it (restarting kernel, etc. does not work, I am using a python 2.7 kernel, it's running on a Mac).

**Comment by takluyver** [comment]
Check `sys.executable` anyway, it might be running the kernel somewhere different to where you expect.

**Comment by horsto** [comment]
ok, `sys.executable` in just python in terminal gives me /Users/hotte/miniconda2/envs/jupyterlab/bin/python 
(after I `source activate` the conda env)
Within the notebook I do the same and it gives me /Users/hotte/miniconda2/bin/python .
Which is pretty puzzling? Any ideas why the notebook itself does not point to my environment that I started it from? (notebook and everything WAS installed from within the environment)

**Comment by takluyver** [comment]
There'll be a kernelspec somewhere which is pointing it to a different Python. We really need to redesign that system. Run `jupyter kernelspec list` in a terminal to see what kernels it knows about. It will point you to directories which contain `kernel.json` files - one of those will be pointing it to the wrong Python.

**Comment by horsto** [comment]
Thanks! I reset the path in the kernel.json file as you suggested. Everything works now!

**Comment by MaxiPython** [comment]
hello, I'm having the same trouble, I followed the instructions from takluyver, but I ended up in:

`$ jupyter kernelspec list`
`Available kernels:`
`  python2    /Users/mcavarra/anaconda/lib/python2.7/site-packages/ipykernel/resources `

then I looked into the directory: 
`cd /Users/mcavarra/anaconda/lib/python2.7/site-packages/ipykernel/resources`
`MassimiianosAir:resources mcavarra$ ls -la`
`total 16`
`drwxr-xr-x   4 mcavarra  staff   136 Jun 29 20:47 .`
`drwxr-xr-x  50 mcavarra  staff  1700 Jul 29 10:40 ..`
`-rw-r--r--   3 mcavarra  staff  1084 May  8  2015 logo-32x32.png`
`-rw-r--r--   3 mcavarra  staff  2180 May  8  2015 logo-64x64.png`
there is no `kernel.json` file there, just 2 logos
can anyone help please ?

**Comment by MaxiPython** [comment]
the package I need (numpy) is already installed by the way:

`$ '/Users/mcavarra/anaconda/envs/gl-env/bin/python' -m pip install numpy`
`Requirement already satisfied (use --upgrade to upgrade): numpy in`
`/Users/mcavarra/anaconda/envs/gl-env/lib/python2.7/site-packages`

**Comment by Vaishak** [comment]
How to reset the path in the kernel.JSON file? 

{
 "argv": [
  "python",
  "-m",
  "ipykernel_launcher",
  "-f",
  "{connection_file}"
 ],
 "display_name": "Python 3",
 "language": "python"
}

the path to where my sys.executable points to at system: 
`/home/whyshock/anaconda3/bin/python`
the path to where my sys.executable points to in the virtualenv named plotly:
`/home/whyshock/vs-workspace/plotly/bin/python`

When I run on terminal
jupyter kernelspec list 
`  python3    /home/whyshock/anaconda3/share/jupyter/kernels/python3`

I'm trying to run the jupyter notebook in the virtual environment when I try to run a program I get a modulenotfoundError but if I try to pip install the module, it says requirement already satisfied!

What am I doing wrong and how can I fix this?

**Comment by takluyver** [comment]
Where it says `"python"` in that file, give it the path to the Python executable you want it to use.

**Comment by tariqueye** [comment]
@takluyver Thanks for the blog post!

**Comment by mmagnuski** [comment]
@takluyver 
I'm having the same problem but sys.executable is the same in notebook and console. Below you can see that I can't import a package (installed with `python setup.py develop`) in the notebook but am able to do so from the console.
![image](https://user-images.githubusercontent.com/8452354/33241485-51cfbf56-d2c6-11e7-9564-2f174ce254f7.png)
(I've been restarting notebook kernel many times)

**Comment by takluyver** [comment]
Do `print(mypy)` in the console to see where it's imported from. Is it in the working directory? In that case, it will only be picked up when running in that directory. Otherwise, check `sys.path` for differences.

**Comment by mmagnuski** [comment]
The package is not in the working directory. I can see the differences in `sys.path` but I still don't understand why the same python environment started from the same console has the path to this package or not depending on whether it is run via jupyter notebook. :(

**Comment by tariqueye** [comment]
Having this issue bites...

However this post can help, I'm on a Mac and viola it worked.  In a nut
shell run sys.path in your terminal to locate all your python installs
including those in Anaconda.  Then locate the package you installed and see
what directory it was installed in.  Then go into your Jupyter notebook and
run sys.path to find where your notebook is installed.  If your having
issues with the package you should clearly see they are not in the same
directories.

Follow the steps in http://takluyver.github.io/posts/i-cant-import-it.html to
remedy the issue.

Its frustrating but you'll figure it out.

[image: photo]
*Tarique Hasheem*
510-299-2680
<http://www.linkedin.com/in/tariqueh>
Create your own email signature
<https://wisestamp.com/email-install?utm_source=promotion&utm_medium=signature&utm_campaign=create_your_own>

On Sun, Nov 26, 2017 at 10:37 AM, Mikolaj Magnuski <notifications@github.com

**Comment by tariqueye** [comment]
Regarding your last email....

You have to be in your conda env (source activate name_of_your_conda_env).
Then you must install the package with conda install name_of_package.

Otherwise using pip or pip3 will install in a directory you may not want.
If the package is not available via conda install you can go into your py
directory where the conda env is and us pip but make sure to provide the
entire path to the directory for the install to be successful.


[image: photo]
*Tarique Hasheem*
510-299-2680
<http://www.linkedin.com/in/tariqueh>
Create your own email signature
<https://wisestamp.com/email-install?utm_source=promotion&utm_medium=signature&utm_campaign=create_your_own>

**Comment by mmagnuski** [comment]
@tariqueye 
I have to install some packages from github and still be able to edit so I use `python setup.py develop`. I can add a line of code to append a relevant path to the sys.path, I just found it surprising that I can import from the console and not from the notebook although the same python env is used in both.

**Comment by gzewdie** [comment]
@all,
I did the same issue when try to import seaborn. I fixed it by installing seaborn as sudo on ubuntu!
sudo pip install seaborn 

Then I could import it on jupyter notebook. Hope this helps!

**Comment by goyalritesh21** [comment]
@fonnesbeck I also had the same issue 
![pic](https://user-images.githubusercontent.com/22390990/35474389-a8e8be14-03b3-11e8-9a22-d1ed646e468a.png)

Then I checked the list of packages in the conda virtual environment.
You can check the packages in your virtual environment using the following command:
* If the environment is not activated, in your Terminal window or an Anaconda Prompt, run:
        `conda list -n env_name`

* If the environment is activated, in your Terminal window or an Anaconda Prompt, run:
        `conda list`

When we make a virtual environment, all the packages are not inherited. Only sone of them are inherited from parent directory. To resolve this problem, you can use the following command to install the missing package:
         `conda install package_name`

This will install the package on your environment and your jupyter notebook will not give error.

**Comment by mmagnuski** [comment]
@all
For some reason it seems to work on that problematic machine now. 😄 I can't identify what helped.

**Comment by cateoliveira** [comment]
To install on notebook and have any effect the correct way to do it is:

import sys
!conda install --yes --prefix {sys.prefix} seaborn

It will install in all environments.

**Comment by Maier123** [comment]
@ALL 
I'm having the same issue. I am running python 3.6 on a windows 8 computer and write my code with Notepad++, so i couldn't refer to the conda or jupyter answers. 
my sys.executble gives me 'C:\\Users\\MyName\\AppData\\Local\\Pprograms\\Python\\Python36-32\\python.exe'
sys.path gives me 'C:\\Users\\....\\Pyhton36-32\\lib\\site-packages\\pyinsane2-2.0.2-py3.6-win32.egg' with pyinsane2 is the module I want to use, which is installed and named as "requirement already satisfied " when running >>>import pyinsane2. (I couldn't tell why it is not the pyinsane2 directory in \site-packages instead of the pyinsane2-2.0.2-py3.6-win32.egg )
How do I get python.exe and my package into the same directory?? Thanks

**Comment by smeeklai** [comment]
@Maier123 Are you running a notebook on conda virtual env?

**Comment by Maier123** [comment]
I'm not sure what that means but I only installed Pyhton 3.6 and Notepad++ on my computer to code. Also, My pyhton is added to the path.

**Comment by smeeklai** [comment]
@Maier123 How about trying to follow [these steps](http://jupyter.readthedocs.io/en/latest/install.html) first, then creating [conda](https://conda.io/docs/user-guide/tasks/manage-environments.html#creating-an-environment-with-commands) environment. After that, activate the environment, `pip install ...` whatever you need, and then run `jupyter notebook`

**Comment by lyupin** [comment]
conda install nb_conda_kernels
I tried this. It worked. Original post: https://stackoverflow.com/a/44786736/4110004

**Comment by smilerz** [comment]
I was having a similar issue, I install tflearn under a conda environment but was getting 'module not found' when attempting to import it even though I had run jupyter from the correct conda environment.

However, I had _not installed jupyter in that environment_, so it was running from the default installation.
Installing jupyter into the target environment corrected the problem.

**Comment by mehemken** [comment]
# For what it's worth

tl;dr I couln't get `import pymc3 as pm` to work on Jupyter Lab so I had to resort to notebook. In the grand scheme of things, I think I'll be ok.

## The missing piece

The last line of the traceback from `import pymc3 as pm` is 
```
RuntimeError: To use MKL 2018 with Theano you MUST set "MKL_THREADING_LAYER=GNU" in your environement.
```
Thus
```
export MKL_THREADING_LAYER=GNU
```
seemed to fix `python`, `notebook` and `ipython`, but not `lab`.

## Results

### Python

This works. Sure there's a warning, but the import actually imports `pymc3`.
```
$ python
Python 3.6.5 |Anaconda, Inc.| (default, Apr 29 2018, 16:14:56) 
[GCC 7.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pymc3 as pm
/home/emilio/anaconda3/envs/bmh/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
```

### Jupyter notebook

This gives the same warning, but it works.
```
$ jupyter notebook
# In a cell
In [1] import pymc3 as pm
/home/emilio/anaconda3/envs/bmh/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
```

### ipython

This gives the same warning as above, but it works.
```
$ ipython
Python 3.6.5 |Anaconda, Inc.| (default, Apr 29 2018, 16:14:56) 
Type 'copyright', 'credits' or 'license' for more information
IPython 6.3.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import pymc3 as pm
```

### Jupyter lab

This fails
```
$ jupyer lab
# In a cell
In [1] import pymc3 as pm
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
<ipython-input-1-2e5c536b7c65> in <module>()
----> 1 import pymc3 as pm

ModuleNotFoundError: No module named 'pymc3'
```

## My system

- Ubuntu 18.04
- conda 4.5.2
- jupyter 4.4.0
- jupyter notebook 5.4.1
- jupyter lab 0.31.5
- ipython 6.3.1
- python 3.6.5
- pymc3 3.3

## Conclusion

I've spent enough time on this so I'm just going to use the notebook. I'd love to use Jupyter Lab, but there's a limit to the time I can allocate to troubleshooting environments.

**Comment by soumyapatro** [comment]
I faced the same issue too. What was happening was that my ipython notebook was using the global instance of notebook installation since my newly created conda environment did not have jupyter notebook installed in it. And it was also searching for all the packages referenced in the notebook globally.

The fix was : conda install jupyter-notebook in the newly created environment.

After doing this, when I launched ipynb, it started using the packages in my environment and not the global ones. I also used conda install command for installing all the packages. 

To test whether it is a global/environment issue export your ipynb as .python file and try running it from bash. If it works in the environment's bash, you know that your notebook is trying to find packages in the global env.

**Comment by vtslab** [comment]
I also had the problem of differences in the `sys.executable` for the console and the jupyter notebook.
Both point to different pythons within the same virtual environment.
My solution was to install jupyter notebook in the virtual environment
`source activate <your virtual env>`
`conda install jupyter notebook`
After killing and restarting jupyter notebook they are the same :-)

**Comment by iamjli** [comment]
Despite all the helpful comments, it took me a while to figure this one out. I'm using pip, not conda, and I got it to work reliably without having to manually change the Python executable path as follows: 

1) Enter virtual environment: `source venv/bin/activate`
2) Make sure to reinstall jupyter: `pip3 install jupyter`
3) Restart venv by deactivating and reactivating.

Checking `sys.executable` should now point to the venv's Python kernel. I prefer this method over manually changing `kernel.json` since that would get tedious if working with multiple different venvs.

**Comment by amandalmia14** [comment]
Install jupyter in that particular environment.

**Comment by jadianes** [comment]
Thanks @iamjli ! The bit I was missing was actually re-activating the environment after `pip3 install jupyter`.

**Comment by jimht011** [comment]
I'm pretty much sure the seaborn 9.0 conda installation package has issue. I got same error after install seaborn 9.0. I tried seaborn conda 8.1 and it is working. I don't have issues with other conda installation packages.

**Comment by sandsece** [comment]
Run the following in the jupyter notebook cell:

import sys

sys.path

sys.executable

It may not be pointing to your virtual environment but to the root

The fix is to install the jupyter notebook from inside your virtual environment

$ . your_env/bin/activiate

(your_env)$ python -m pip install jupyter

Now you can import tensorflow or keras

**Comment by shiningbridge** [comment]
Thanks all for the infomatic post for the issue that I also encountered. And agree with @sandsece 
i.e. 

==========
### PROBLEM:

From within myENV created from conda, run jupyter notebook, but in this notebook some module that could be loaded from myENV just through out error like:
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
<ipython-input-6-ae142f98063f> in <module>()
      1 # import libraries
----> 2 import folium

ModuleNotFoundError: No module named 'folium'


===========
### Check:

1. `which jupyter`
since even sys.path show correct path of myENV, the sys.executable can be different in myENV and in Jupyter notebook. This can simply because the jupyter you run is not installed in myENV and it has to fallback to the base install. It happens that in the base environment the module you want to use is not installed. ----> error!!!

===========
### Solution: 
For the case above, it is straightforward:
`conda install jupyter`


Hope this will help. 

BR//Kangqiao

**Comment by auwsom** [comment]
For me, I installed Jupyter after creating an environment, but then was trying to run a module installed from the base env. I found by "jupyter kernelspec list" (https://github.com/jupyter/notebook/issues/2563), my kernel.json at C:\Users\username\Anaconda37\share\jupyter\kernels\python3\kernel.json was pointing to the python.exe in my working env. Changed the path and solved it. 

[This ](https://stackoverflow.com/questions/897792/where-is-pythons-sys-path-initialized-from)was an exhaustive description of python path setting. And [here](https://stackoverflow.com/questions/46634660/jupyter-notebook-wrong-sys-path-and-sys-executable) is my solution.

**Comment by jmvictor5656** [comment]
I was unable to import many external packages but when I updated pip to latest version then It started workingfor me I just updated pip and

**Comment by swapnaruvva143** [comment]
> Complete reinstallation of Anaconda fixed the issue.

can you please suggest the better way to solve this issue

**Comment by qifengzhou** [comment]
My problem solved by install Jupyternotebook in the conda environment and use it there. The packages can be imported in the same environment.

Without this step, I think you are using Jupyternotebook from the general setting, for that import packages in a specific conda environment is not working.

**Comment by alkair** [comment]
Hey, 
I just had the same problem and it was a stupid mistake. I have cloned the environment and started it. After opening jupyter notebook it took the Kernel from previous session. It happened too be that it had similar name since it comes from similar environment. well...always double check before you blame and say bad things to your PC :D

**Comment by akanshajainn** [comment]
> There'll be a kernelspec somewhere which is pointing it to a different Python. We really need to redesign that system. Run `jupyter kernelspec list` in a terminal to see what kernels it knows about. It will point you to directories which contain `kernel.json` files - one of those will be pointing it to the wrong Python.

What if everything is fine as everyone's told here but still it doesn't work ?

**Comment by mutiev** [comment]
Ok, i solved this problem in my case:
Windows 10
Anaconda 3
installed tensorflow venv

When I use Python3 kernel - it works fine. 
When I switched to tf-gpu venv - it's fires "No module named 'seaborn'

But in previous cases, when I install other modules via jupyter in cell (!conda install <package> or !pip install <package>) everything worked ok.
But in case with seaborn it didn't
So I opened terminal > activate tf-gpu > conda install seaborn 
And problem were fixed

**Comment by FarnamAdelkhani** [comment]
I had no issues with install on my MacBook, however, on my Windows 10 64-System, I had the same problem. I resolved it as such:

Try running anaconda prompt as 'administrator', then use:
> conda install -c conda-forge pydotplus

**Comment by ponugoti** [comment]
On Ubuntu, I found that using `sudo ~/anaconda3/bin/pip` instead of `pip` or `pip3` worked for me.

Additionally, I've just started using an alias for pip for convenience with `alias pip="sudo ~/anaconda3/bin/pip` in the terminal.

**Comment by EkateKK** [comment]
> On Ubuntu, I found that using `sudo ~/anaconda3/bin/pip` instead of `pip` or `pip3` worked for me.
> 
> Additionally, I've just started using an alias for pip for convenience with `alias pip="sudo ~/anaconda3/bin/pip"` in the terminal.

this worked for me as well

**Comment by 1998anwesha** [comment]
how to import apriori package in jupyter? I have already installed it in anaconda prompt but jupyter does not import it

**Comment by amandalmia14** [comment]
@1998anwesha 
Just check where is the Python Interpreter is running? Are they same or different?
Both Jupyter and your interpreter must have same interpreter.

`
import sys
print(sys.executable)
`

**Comment by 1998anwesha** [comment]
@amandalmia14 Thank you

**Comment by Abhishek-31** [comment]
I was having the same issue and as our system contains many instances of python, many times jupyter notebook starts using some other version and hence we can't do anything but solve the issue by bare hands by changing the package.json file present in the path given by 
    `sys.executable`

**Comment by amandalmia14** [comment]
I would recommend everyone to use different virtual environment for every project or different environments in Anaconda. (Will insist on using Anaconda because of its great library  version management)

**Comment by FredrikLastow** [comment]
I was having the same issue where my notebook couldn't import certain modules, but a python console in the terminal had no problems. I fixed it by creating a new kernel for my conda environment ([see this link](https://ipython.readthedocs.io/en/stable/install/kernel_install.html#kernels-for-different-environments)) by running:
~~~~ 
source activate myenv
python -m ipykernel install --user --name myenv --display-name "Python (myenv)"
~~~~
Now, if I created a new notebook from my newly created kernel I could import all my modules without a hitch. My `sys.executable` and `sys.path` are now the same in the terminal and my notebook. Hope this helps!

**Comment by awa5114** [comment]
Well if it will only work if we install a new `jupyter notebook` in each new environment, what on earth is the point of having these stupid `kernelspec` files and separate kernels in the base jupyter notebook installation? Also the recommended way of handling multiple environments according to the docs is to use kernels and kernelspecs. But this doesn't seem to work, instead everyone is doing this workaround of installing a new jupyter notebook on each environment.  Ridiculous!

**Comment by FredrikLastow** [comment]
@awa5114 I installed nb_conda_kernels in my base conda environment which did what the expected behaviour is of the kernelspecmanager by installing a new one. This fixed the issue for me and now each conda environments has its own kernel. I can now choose which kernel aka conda environment my notebooks uses. I'm not sure if we are using the same instance of Jupyter though. I'm using Jupyterhub, but maybe this make much of a difference in this case. Hope this helps!

**Comment by ChaosPredictor** [comment]
Same issue here, just installed sklearn by conda, by: `conda install scikit-leanr`
I can import it from command line, but not from jupyter.

**Comment by mytrjp** [comment]
same here. 
what resolved my problem is that I first checked available kernels by 'jupyter kernelspec list'
This listed me available kernels, but I noticed that it was using another kernel that I am not in
Then I removed that kernel completely and it was resolved.

**Comment by aseem-hegshetye** [comment]
I had the same problem. Changing kernels did not work for me.
 I installed pandas in my venv but It wont show up in jupyter which I ran through my venv.

I have Ubuntu, with a virtual env in python 3.7 
`python3.7 -m venv my_env`
Then I added the venv python site-packages folder to system path and it worked. Added following code to jupyter notebook and it was able to find pandas easily

```
import sys
sys.path.append('/home/aseem/venv/general_venv/lib/python3.7/site-packages')
```

**Comment by lumbric** [comment]
I've tried summarizing options and common troubles in this Q&A:
https://stackoverflow.com/questions/58068818/how-to-use-jupyter-notebooks-in-a-conda-environment

**Comment by the-c0d3r** [comment]
This is quite a weird bug. Like @iamjli mentioned, if I just reactivate the env by deactivating and activating again, it works fine, even though literally nothing else changed. 
So that means if you install package and jupyter at the same time, without reactivating env, it will not be able to import, until you reactivate. I just tested this again in a new virtualenv, it proved to be correct.

```bash
mkdir -p /tmp/test
cd /tmp/test
virtualenv -p python3 .venv
source .venv/bin/activate
pip install numpy && pip install jupyter && jupyter notebook
```

create a new notebook and try to import numpy. It **failed**.
Try deactivate, and activate again, it works fine. 
I have a hard time believing such fundamental issue has existed since 2017 and remain unfixed.

**Comment by chrisPiemonte** [comment]
My problem was that jupyter (jupyter-lab) was not installed in the environment i was using. I installed jupyter lab in that conda environment and newly installed packages were imported correctly

**Comment by penestia** [comment]
> I had the same problem. Changing kernels did not work for me.
> I installed pandas in my venv but It wont show up in jupyter which I ran through my venv.
> 
> I have Ubuntu, with a virtual env in python 3.7
> `python3.7 -m venv my_env`
> Then I added the venv python site-packages folder to system path and it worked. Added following code to jupyter notebook and it was able to find pandas easily
> 
> ```
> import sys
> sys.path.append('/home/aseem/venv/general_venv/lib/python3.7/site-packages')
> ```

This worked for me and seems like a very easy solution.

**Comment by shadzic** [comment]
I had the same issue.
Changing the kernel.json file did not work for me.
I noticed that sys.executable was different between my terminal and the notebook. The notebook did not take the environment into account.
I followed this: 
https://stackoverflow.com/questions/58645807/change-interpreter-in-jupyter-notebook
And it solved the issue

**Comment by alexcadillon** [comment]
I did `conda list` in both my `base` and my `new_env` and noticed that both `jupyterlab` and `jupyterlab_server` were missing in `new_env`.
So basically I just did `conda install -c conda-forge jupyterlab` while having `new_env` activated and now everything runs well.

I thought to look into this because `new_env` was being used whenever I ran my notebook on JupyterNotebook but not when I ran it on JupyterLab.

**Comment by ShayekhBinIslam** [comment]
If u are using conda;
1. type: jupyter kernelspec list
   inside jupyter notebook
2. open kernel.json file for python
3. edit and paste
```
{
 "argv": ["python3", "-m", "IPython.kernel",
          "-f", "{connection_file}"],
 "display_name": "Python 3",
 "language": "python"
}
```
Reference: https://jupyter-client.readthedocs.io/en/stable/kernels.html

**Comment by ShayekhBinIslam** [comment]
> 
> 
> How to reset the path in the kernel.JSON file?
> 
> ```json
> {
>  "argv": [
>   "python",
>   "-m",
>   "ipykernel_launcher",
>   "-f",
>   "{connection_file}"
>  ],
>  "display_name": "Python 3",
>  "language": "python"
> }
> ```
> 
> the path to where my sys.executable points to at system:
> `/home/whyshock/anaconda3/bin/python`
> the path to where my sys.executable points to in the virtualenv named plotly:
> `/home/whyshock/vs-workspace/plotly/bin/python`
> 
> When I run on terminal
> jupyter kernelspec list
> ` python3 /home/whyshock/anaconda3/share/jupyter/kernels/python3`
> 
> I'm trying to run the jupyter notebook in the virtual environment when I try to run a program I get a modulenotfoundError but if I try to pip install the module, it says requirement already satisfied!
> 
> What am I doing wrong and how can I fix this?

replace
 "argv": [
>   "python",
by

 "argv": [
>   "python3",

**Comment by tiru-patel** [comment]
> Wow, that's bizarre. Its an extremely odd list. I've deleted `~/Library/Jupyter/kernels` and that appears to have straightened things out.
> 
> Thanks. It worked

**Comment by RydellMountain** [comment]
Hi people. So what i did was i actually ran 

> pip install seaborn

In Jupyter notebook,after restarting the kernel i imported seaborn, and it worked

**Comment by jmason86** [comment]
Same problem, for me the package is `mpld3`. Installed fresh conda environment from a yml, added the kernel to jupyter with 
```
python -m ipykernel install --user --name=helioml_ch2
```
Import of `mpld3` works in ipython command line but not jupyter. It seems that even if there is a workaround, this issue should still be open because the expected behavior does not occur.

**Comment by jmason86** [comment]
Debugged the issue, at least for my case. (The workaround of deleting everything in ~/Library/Jupyter/kernels didn't do anything for me.)

This doesn't work: 
```
conda env create -f environment.yml -n test
python -m ipykernel install --user --name=test
jupyter lab
```
Same issue here: import works when in ipython command line but not in jupyter (notebook or lab). 

This _does_ work: 
```
conda env create -f environmentyml -n test
source activate test
ipython kernel install --user --name=test
```
Haven't tested doing `conda activate test` but imagine that should work too.

**Comment by shantisuresh01** [comment]
Thank you!  Deleting the corresponding Jupyter kernel in ~/Library/Jupyter/kernels/ (or moving it aside, if one is paranoid), worked for imports to start working in Jupyter Lab.  In my case, it was "~/Library/Jupyter/kernels/python3".

**Comment by quents** [comment]
> `conda install nb_conda_kernels`
> I tried this. It worked. Original post: https://stackoverflow.com/a/44786736/4110004

This works! Thank you.

**Comment by NickBenthem** [comment]
The solution for me was to see what the actual python executable was being used:
```
import sys
print(sys.executable)
```

In my case, I had installed my Jupyterhub into `/opt/conda/`, so my pip install was over in `/opt/conda/`. Running 

```
!/opt/conda/bin/pip3 install torch
```

then restarting the kernel worked.

**Comment by cgironda** [comment]
Hi guys,

After all these years this problem is still annoying. I just wanted to quote what I did to solve a similar problem started by @fonnesbeck a time ago.

After checking the comment of @Vaishak I decided to open my `kernel.json` file that is inside the environment that I created for my project:
/anaconda/envs/MYPROJECT/share/jupyter/kernels/python3

and I changed "python" by "/Users/myname/anaconda/envs/MYPROJECT/bin/python"

```
{
 "argv": [
  "/Users/myname/anaconda/envs/MYPROJECT/bin/python",
  "-m",
  "ipykernel_launcher",
  "-f",
  "{connection_file}"
 ],
 "display_name": "Python 3",
 "language": "python"
}
```
This was possible because of the discussion between @ShayekhBinIslam  and @Vaishak

**Comment by kumarapurv** [comment]
I faced a similar issue. While the imports were working fine on terminal in the virtual environment, they failed to import in the jupyter notebook. The problem was that I had not installed a local jupyter package for the virtual environment. Therefore, the jupyter notebook on my newly created virtual environment was using the global instance of notebook installation, i.e. from the base environment.

I used the command  
> conda install jupyter notebook

to install notebook on my virtual environment, and this fixed all issues for me. I could import all the packages local to the newly created virtual environment.

**Comment by saguillo2000** [comment]
> Complete reinstallation of Anaconda fixed the issue.

I fixed the problem by uninstalling Python 3.9 interpreter and only keeping Anaconda's interpreter. The steps I followed are:

1. Uninstall any Python interpreter installed in your OS, except Conda.
2. Execute Conda's enviornment and reinstall Jupyter: pip install jupyter.

This worked for me. In my case ,  I wasn't using the other interpreter. To test if this works, execute sys.executable in Jupyter and it should show your Conda's Python enviornment path.
